docs/proposals/cluster-bootstrap-with-gossip.md

```diff
@@ -0,0 +1,153 @@
+
+In the kubeadm command, in general:
+
+1. If --discovery flag is "gossip", which is the default:
+
+    1. If --token has been provided, remember it.
+
+    2. If --token has not been provided, generate a random string and print it to the user.
+
+    3. Write the following to /etc/kubernetes/kubelet-bootstrap.json:
+
+```
+{
+    "apiVersion": "v1alpha1",
+    "kind": "Discovery",
+    "role": "<role>",
+    "discovery": {
+        "type": "Gossip",
+        "token": "<token>",
+        "peers": []
+    }
+}
+```
+
+2. If --discovery flag is "out-of-band", require --apiserver-urls and --cacert arguments to also be passed.
+
+    4. In this case, disable the gossip server, and go straight into **bootstrapping** mode.
+
+    5. Write the following to /etc/kubernetes/kubelet-bootstrap.json:
+
+```
+{
+    "apiVersion": "v1alpha1",
+    "kind": "Discovery",
+    "role": "<role>",
+    "discovery": {
+        "type": "OutOfBand",
+        "apiserverURLs": ["https://foo/"],
+        "cacertFile": "/path/to/cacert.pem"
+    }
+}
+```
+
+In future, there could be other --discovery implementations, such as DNS- or discovery service-based.
+
+The Golang Discovery interface should be:
+
+```
+type Discovery interface {
+    Start()
+    Discover() (apiServerUrls []string, cacert x509.Certificate, err error)
+}
+```
+
+Start should be called every time the kubelet starts up, and it should return immediately, in case the discovery mechanism needs to run a persistent service (like a gossip socket).
+
+### Kubelet awakens
+
+The kubelet waits in **pending** mode for the bootstrap file to show up. When it does, it unmarshals the Discovery object from JSON depending on the "kind" field, and calls `Start()` and then `Discover()` on it.
+
+When it does, it's the Discovery implementation's job to move into **bootstrapping**.
+
+#### Masters
+
+Initializing a new master shouldn't require TLS bootstrap, because the master already has privileged access to the API server. New masters have to add their own address into the API server URL list in the mesh case.
+
+#### Nodes
+
+For the out-of-band implementation it should be pretty obvious how to satisfy that interface with an implementation. The gossip implementation is slightly more clever...
+
+### Gossip implementation
+
+As soon as has Discover called on it, it attempts to form a secure mesh network using the token and the peers, using [a simple gossip protocol library](https://github.com/weaveworks/mesh). For more information on the library, see [this talk](http://infoq.com/presentations/weave-mesh) by the authors.
+
+When that happens, kubelet uses [CRDTs](https://github.com/weaveworks/kubelet-mesh/blob/master/state.go) to support gossiping CA certs and lists of URLs of API servers. As soon as a kubelet learns of both of these pieces of information, it returns from the Discover method, and kubelet proceeds to attempt to perform TLS bootstrap against the API server running on the master.
```

Comment:
AIUI, if you compromise a node on the mesh and read the mesh key out of memory/disk then you'd be able to read the cacert and api server URLs, no more than that.

Fixed the dead link, sorry about that. The link goes to our trivial CRDT implementation for merging API server URLs and deciding on CA certs deterministically based on their creation dates (falling back to comparing signatures if the dates are equal). This took us half a day to code, based on the sample code here: https://github.com/weaveworks/mesh/tree/master/examples/increment-only-counter and watching http://infoq.com/presentations/weave-mesh